gateway/modules/crud/batcher.go

```diff
@@ -0,0 +1,129 @@
+package crud
+
+import (
+	"context"
+	"time"
+
+	"github.com/spaceuptech/space-cloud/gateway/model"
+	"github.com/spaceuptech/space-cloud/gateway/utils"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spaceuptech/space-cloud/gateway/config"
+)
+
+// data structure for storing a group of channels
+// against particular table/collection of particular database
+type (
+	batchMap map[string]map[string]map[string]batchChannels // keys are project name, dbAlias, table name
+
+	batchChannels struct {
+		request batchRequestChan
+		closeC  chan struct{}
+	}
+
+	batchRequestChan  chan batchRequest
+	batchResponseChan chan batchResponse
+
+	batchRequest struct {
+		document interface{}
+		response batchResponseChan
+	}
+	batchResponse struct {
+		err          error
+		docsInserted int64
+	}
+)
+
+// closeBatchOperation closes all go routines associated with individual collection for batch operation
+func (m *Module) closeBatchOperation() {
+	for _, projectValue := range m.batchMapTableToChan {
+		for _, dbAliasValue := range projectValue {
+			for _, tableValue := range dbAliasValue {
+				tableValue.closeC <- struct{}{}
+			}
+		}
+	}
+}
+
+// initBatchOperation creates go routines for executing batch operation associated with individual collection
+func (m *Module) initBatchOperation(project string, crud config.Crud) {
+	batch := batchMap{}
+	for dbAlias, dbInfo := range crud {
+		if dbInfo.Enabled {
+			for tableName, tableInfo := range dbInfo.Collections {
+				closeC := make(chan struct{})                    // channel for closing go routine
+				addInsertToBatchCh := make(batchRequestChan, 20) // channel for adding request to batch op
+				go m.insertBatchExecutor(closeC, addInsertToBatchCh, dbInfo.BatchTime, project, dbAlias, tableName, tableInfo.BatchRecords)
+				if batch[project] == nil {
+					batch[project] = map[string]map[string]batchChannels{dbAlias: {tableName: {request: addInsertToBatchCh, closeC: closeC}}}
+					continue
+				}
+				batch[project][dbAlias][tableName] = batchChannels{request: addInsertToBatchCh, closeC: closeC}
```

Comment:
What if `batch[project][dbAlias] id nil or that keg doesn't exist? You should check for that as well

Reply:
this condition will not occur
as the map 's value for every project will be nil
and i check for nil, if nill i initialze the complete remaining data structure
check the above if case